cirq/circuits/circuit.py

```diff
@@ -964,6 +965,23 @@ def unitary(
         result = _apply_unitary_circuit(self, state, qs, dtype)
         return result.reshape((side_len, side_len))
 
+    def is_noisy(self) -> bool:
```

Comment:
Quick question: why doesn't `has_unitary` work for you? Can we change that to cater for your requirements? What are those?

Reply:
Measurements and parameters cause `has_unitary` to return false. This is the correct behavior; a circuit with a measurement or an unresolved parameter _cannot_ produce a unitary without ignoring or resolving those elements first.

However, such elements are not noisy - measurement is a permissible non-unitary in noiseless circuits, and parameterized objects with a noiseless base type will resolve to a noiseless object.

Reply:
Even circuits with mid-circuit measurements are considered noiseless?

Reply:
For the purposes of qsim, yes. The base qsim simulator can handle mid-circuit measurements; it can't handle a `Channel`.

More generally, while both measurement and noise can be represented as channels with some set of operators, only measurement reveals which operator was applied. For noise, this is only possible in simulation; based on this distinction, I don't think measurement should be considered "noise" even though it makes a circuit non-unitary.